### PR TITLE
[MLA] Pass vmem_limit_bytes to xpose_pipeline instead of relying on LIBTPU_INIT_ARGS

### PR DESCRIPTION
**Pull request by QiliangCui2023**
## Problem

The MLA v2 kernel's q_nope and output transposes (`xpose_pipeline` pallas_call) need ~33.3–33.6 MiB of scoped VMEM, which exceeds XLA's 32 MiB default (`--xla_tpu_scoped_vmem_limit_kib`) and fails compilation with `E1001: CompileTimeScopedVmemOom` on every cold-cache run. This is the "tpu7x/tpu6e JAX unit tests - kernels" failure on main builds 22345 and 22385 (`mla_v2_test.py::test_ragged_paged_attention_basic`, 33.56M vs 32.00M limit).

The previous mitigation (#3159, before that #3011's conftest setting) appended `--xla_tpu_scoped_vmem_limit_kib=65536` to `LIBTPU_INIT_ARGS` at MLA-test-module import. That is **order-dependent**: under `pytest tests/kernels`, `deepseek_v4/compressor_test.py` evaluates `jax.devices()` at import inside its `skipif` decorator, and `deepseek_v4/` collects alphabetically before `mla_*`, so libtpu initializes **before** the env var is appended and the flag is silently dropped. Whether the kernels CI job passed then depended purely on JAX compilation-cache warmth — warm cache skips the compile (no OOM), cold cache always fails. A 10x A/B/C stress matrix at the pre-#3134, #3134, and #3028 commits showed a flat ~2/10 cold-failure rate at all three, confirming the flake is environmental and none of those PRs caused it.

## Fix

`mla_ragged_paged_attention` already resolves `vmem_limit_bytes` (`None` → full VMEM capacity via `pltpu.get_tpu_info().vmem_capacity_bytes`, [kernel.py#L2268-L2269](https://github.com/vllm-project/tpu-inference/blob/589c3f9ea2b68c598bc5b4a7e2017b1184a0e11e/tpu_inference/kernels/mla/v2/kernel.py#L2268-L2269) — 64 MiB on tpu7x) and applies it to the main pallas_call through Mosaic `CompilerParams`. This PR threads the same value through `prepare_q_nope_inputs` / `prepare_outputs` into `xpose_pipeline`'s `CompilerParams`. The per-kernel Mosaic limit overrides the global XLA flag for that custom call only, so:

- no environment variable is needed at all (the fragile `LIBTPU_INIT_ARGS` blocks in `mla_v2_test.py` / `mla_tuned_vs_baseline_test.py` are removed), and
- nothing changes for any other compiled program (the #3159 concern — a global flag altering XLA codegen for unrelated e2e tests — cannot recur).

Both OOM-ing call sites are covered: `prepare_outputs` (384x128x512 bf16, 33.56M) and `prepare_q_nope_inputs` (128x352x512 f32, 33.30M).

## Verification

On a tpu7x-8 host, libtpu 0.0.43, with the JAX compilation cache disabled (`JAX_ENABLE_COMPILATION_CACHE=false`) to force cold compiles, over `tests/kernels/{mla_v2_test,mla_tuned_vs_baseline_test,transpose_test,deepseek_v4}`:

| | result |
|---|---|
| before, no flag | **8 failed** (all `CompileTimeScopedVmemOom` in `xpose_pipeline`: basic, basic_with_small_page_size, deepseekv3_batch_decode, mixed, prefill_only, sliding_window0/1/2), 173 passed |
| after (this PR) | **182 passed, 0 failed** |

A regression test is added in `transpose_test.py` with the exact failing shape (384x128x512 bf16, n_tile=128, m_tile=64) and `vmem_limit_bytes` set.

Follow-up (separate change): `deepseek_v4/compressor_test.py`'s import-time `jax.devices()` remains a collection-order trap for any future module-level env configuration; it should move to a lazily-evaluated skip condition.

### CODE REVIEW

**Review by gxd3 — APPROVED** [review]

**Comment by QiliangCui2023** [comment]
## Test evidence

All local runs on a tpu7x-8 host, libtpu 0.0.43 (via `TPU_LIBRARY_PATH`), with the persistent compilation cache hard-disabled (`JAX_ENABLE_COMPILATION_CACHE=false`, `JAX_COMPILATION_CACHE_DIR` unset) so every compile is real and warm-cache masking is impossible.

**1. Baseline A (pre-fix code, `LIBTPU_INIT_ARGS` hacks removed, cold):**

```
pytest -q tests/kernels/mla_v2_test.py tests/kernels/mla_tuned_vs_baseline_test.py \
          tests/kernels/transpose_test.py tests/kernels/deepseek_v4/
=> 8 failed, 173 passed in 1196.79s (0:19:56)
```

All 8 failures are `E1001 CompileTimeScopedVmemOom` in `jit(mla_ragged_paged_attention)/jit(xpose_pipeline)`:
`test_ragged_paged_attention_basic`, `basic_with_small_page_size`, `deepseekv3_batch_decode`, `mixed`, `prefill_only`, `sliding_window0/1/2` — covering both call sites:
- `prepare_outputs`: shape 384x128x512 bf16, scoped alloc **33.56M** vs 32.00M limit
- `prepare_q_nope_inputs`: shape 128x352x512 f32, scoped alloc **33.30M** vs 32.00M limit

**2. Verification B (this PR applied — identical command, identical env, only the code change):**

```
=> 182 passed, 0 failed in 1310.55s (0:21:50)
```

The 8 previous failures pass, plus the new regression test. The ~22-min runtime confirms real cold compiles (a warm run of the mla portion takes seconds). A/B is strict: same machine, same libtpu, same pytest invocation; the only variable is the code change.

**3. Regression test** (`transpose_test.py::test_vmem_limit_bytes_allows_large_tiles`): compiles the exact failing shape (384x128x512 bf16, n_tile=128, m_tile=64) with `vmem_limit_bytes=64 MiB` and checks numerical equality against `jnp.transpose`. Compiled cold and passed in run B.

**4. CI build [22497](https://buildkite.com/tpu-commons/tpu-inference-ci/builds/22497) — passed.** The tpu7x kernels job's single first-attempt failure was `test_tuned_vs_baseline` (perf assertion, one -9.18% speedup sample) with **no OOM anywhere in the log** and 1044 tests passed before the `-x` stop; the retry passed the full suite. That perf flake is pre-existing and unrelated: in a 10x stress matrix on unmodified main-history commits it failed 3/30 jobs with the same signature.

**Effective limit note:** with callers passing `vmem_limit_bytes=None`, the kernel resolves it to `pltpu.get_tpu_info().vmem_capacity_bytes` = 64 MiB on tpu7x — byte-identical to the value the old `--xla_tpu_scoped_vmem_limit_kib=65536` flag set when it worked, so no new codegen behavior is introduced for previously-passing paths; the limit is just applied per-custom-call instead of process-globally.

**Not covered:** transpose kernel perf was not re-measured (compiled output should match the old flag-applied path per the note above); no local MLA model serve — CI e2e jobs cover that indirectly.

**Comment by QiliangCui2023** [comment]
**Addendum — import-order + cold-start verified directly on this branch:**

The pre-fix failure was order-dependent (libtpu initializing during collection of `deepseek_v4/compressor_test.py`, before the mla test modules could set `LIBTPU_INIT_ARGS`) and only manifested on cold compiles. Two additional runs on this branch, compilation cache hard-disabled:

1. **CI's exact directory form** (`pytest tests/kernels --ignore=... -k "...basic..."`), where `deepseek_v4/` collects before `mla_*`. A collection-hook trace confirms the worst case occurs — TPU backend already initialized before `mla_v2_test.py` is imported, and no scoped-vmem flag present in `LIBTPU_INIT_ARGS` at any point — and the previously-failing test **passes with a real cold compile** (35.9s). This exact configuration failed deterministically (5/5) before the fix.
2. **Minimal two-file pair with the bad import order** (`compressor_test.py` before `mla_v2_test.py`), the smallest reliable pre-fix repro: **passes cold** (35.5s).

Order independence is structural: the limit now travels inside the traced program (Mosaic `CompilerParams` on the pallas custom call) rather than through a process env var consumed once at libtpu load.

**Review by BirdsOfAFthr — APPROVED** [review]

**Review by kyuyeunk — APPROVED** [review]
